exp/query/api.go

```diff
@@ -315,30 +319,108 @@ func (a *API) vectorSearch(
+			// for a tenant we have locally
+			proceedWithDownload = true
+		} else {
+			// TODO we only care about tenants we have for new versions, make diff func instead of
+			// using bool or return error or other?
+			return nil, "", errors.New("tenant does not exist locally, you can ignore this error, the tenant will be downloaded later")
+		}
+	}
+	if !doUpdateTenantIfExistsLocally {
+		if tenantIsCachedLocally {
+			// tenant is cached, we can just return it because this is not a new version call
+			cachedLsmkvStore := cachedTenantLsmkvStore.(tenantLsmkvStore)
+			return cachedLsmkvStore.s, path.Join(cachedLsmkvStore.localTenantPath, defaultLSMRoot), nil
+		} else {
+			// we should download because the caller wants this tenant to exist locally but it does not
+			proceedWithDownload = true
+		}
+	}
+
+	// base collection path
+	localBaseCollectionPath := path.Join(
+		a.offload.DataPath,
+		strings.ToLower(collection),
+	)
+	// the path we will download to
+	localTenantTimePath := path.Join(
+		localBaseCollectionPath,
+		strings.ToLower(tenant),
+		fmt.Sprintf("%d", currentLocalTime),
+	)
+	// the path we will return
+	localLsmPath := path.Join(
+		localTenantTimePath,
+		defaultLSMRoot,
+	)
+	if proceedWithDownload {
 		// src - s3://<collection>/<tenant>/<node>/
-		// dst (local) - <data-path/<collection>/<tenant>
-		if err := a.offload.Download(ctx, collection, tenant, nodeName); err != nil {
+		// dst (local) - <data-path/<collection>/<tenant>/timestamp
+		if err := a.offload.DownloadToPath(ctx, collection, tenant, nodeName, localTenantTimePath); err != nil {
 			return nil, "", err
 		}
 	}
 
-	// TODO(kavi): Avoid creating store every time?
-	store, err := lsmkv.New(localPath, localPath, a.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
+	store, err := lsmkv.New(localLsmPath, localLsmPath, a.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
 	if err != nil {
 		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
 	}
-	return store, localPath, nil
+	// remember to store the tenant root dir in tenantLsmkvStore, not the lsm path
+	if oldTenantLsmkvStoreAny, ok := a.cachedTenantLsmkvStores.Swap(tenant, tenantLsmkvStore{s: store, localTenantPath: localTenantTimePath}); ok {
+		// when we replace or evict an lsmkvStore, we need to shut it down.
+		// I'm assuming lsmkvStore has internal locks which prevent this shutdown call
+		// from cancelling searches running at the same time.
+		// I'm also assuming once Shutdown returns we don't need the old files on disk anymore.
+		oldTenantLsmkvStore := oldTenantLsmkvStoreAny.(tenantLsmkvStore)
+		oldTenantLsmkvStore.s.Shutdown(ctx)
+
+		// to be extra safe, only remove if the path looks reasonably like what we expect (datapath/collection/tenant/timestamp/*)
+		re := regexp.MustCompile(fmt.Sprintf(`%s/%s/\d+`, localBaseCollectionPath, tenant))
```

Comment:
regexp.Compile is expensive btw. Can we compile once as package variable and use it everywhere?

Reply:
This regex is per collection/tenant so as is we can't just compile it as a package var. However, I've added a comment that if the regex is too slow we should replace it with a prefix match (dropping the timestamp check)